### PR TITLE
SvgSurface: make filename param optional

### LINKED ISSUE #293: SvgSurface path should be an Option
Per the docs in https://www.cairographics.org/manual/cairo-SVG-Surfaces.html the filename param of  `cairo::SvgSurface::new` should be optional see

> filename | a filename for the SVG output (must be writable), NULL may be used to specify no output. This will generate a SVG surface that may be queried and used as a source, without generating a temporary file.

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Per the docs https://www.cairographics.org/manual/cairo-SVG-Surfaces.html
This allows drawing parts of a SVG file without saving it to a tmp file.

I'm not sure if what I did is right or not :P 

Fixes #293

### CODE REVIEW

**Comment by bilelmoussaoui** [comment]
@GuillaumeGomez Any idea why the CI fails with 
```
error[E0599]: no method named `to_glib_none` found for type `std::option::Option<&std::string::String>` in the current scope

  --> src/svg.rs:44:61

   |

44 |                 ffi::cairo_svg_surface_create(path.as_ref().to_glib_none().0, width, height),

   |                                                             ^^^^^^^^^^^^ method not found in `std::option::Option<&std::string::String>`

```
It builds and works fine locally here.

**Comment by GuillaumeGomez** [comment]
Try to change your code as follow:

```rust
path.map(|s| s.as_str()).to_glib_none().0
```

**Comment by bilelmoussaoui** [comment]
My bad! I was running the tests without `--features svg`. It should pass this time. Thanks!

**Comment by bilelmoussaoui** [comment]
It fails because tests are run without `--features use_glib` :man_facepalming:

**Comment by GuillaumeGomez** [comment]
Oh then it depends: is it supposed to be run even without glib or not? If so, please don't use `to_glib` calls.

**Comment by bilelmoussaoui** [comment]
It doesn't need to. In such case, what I'm supposed to do?

**Comment by GuillaumeGomez** [comment]
Like this:

```rust
let c_str = CString::new(c_str).unwrap();
// then:
c_str.as_ptr();
```

File: src/svg.rs [hunk 1]
```diff
@@ -33,13 +33,15 @@ impl SvgVersion {
...
-    pub fn new<P: AsRef<Path>>(width: f64, height: f64, path: P) -> Result<SvgSurface, Status> {
-        let path = path.as_ref().to_string_lossy().into_owned();
-        let path = CString::new(path).unwrap();
-
+    pub fn new<P: AsRef<Path>>(
+        width: f64,
+        height: f64,
+        path: Option<P>,
+    ) -> Result<SvgSurface, Status> {
+        let path = path.map(|p| p.as_ref().to_string_lossy().into_owned());
```
Comment (sdroege) [inline src/svg.rs]:
```suggestion
        let path = path.map(|p| CString::new(path.as_ref().as_os_str().as_bytes()));
        let path = path.map(|p| p.as_ptr()).unwrap_or(ptr::null());
```

Ideally you'd do more than that though (Windows-specific special case...): https://github.com/gtk-rs/glib/blob/faff6573602daec764584030e959f7d28a0aef81/src/translate.rs#L421-L457

File: src/svg.rs [hunk 2]
```diff
@@ -2,19 +2,21 @@
...
+use enums::{Status, SurfaceType, SvgVersion};
+use ffi;
 use std::convert::TryFrom;
 use std::ffi::{CStr, CString};
 use std::fmt;
 use std::io;
 use std::mem;
 use std::ops::Deref;
+#[cfg(not(windows))]
+use std::os::unix::prelude::*;
```
Comment (sdroege) [inline src/svg.rs]: Without a windows case this is not going to compile on Windows. Looks ok for UNIX though.

**Comment by sdroege** [comment]
lgtm

File: src/svg.rs [hunk 3]
```diff
@@ -33,13 +35,37 @@ impl SvgVersion {
...
+            let path = path.map(|p| {
+                CString::new(p.as_ref().as_os_str().as_bytes())
+                    .expect("Invalid path with NULL bytes")
+            });
+            path.map(|p| p.as_ptr()).unwrap_or(ptr::null())
+        };
+        #[cfg(windows)]
+        let path = {
+            // Based on GLib path to C
+            let path_str = path
```
Comment (sdroege) [inline src/svg.rs]: This assumes that `path` is not an `Option` and already a `Path`. Needs the `map()`, `as_ref()` part from above too

File: src/svg.rs [hunk 4]
```diff
@@ -33,13 +35,39 @@ impl SvgVersion {
...
+        #[cfg(windows)]
+        let path = {
+            // Based on GLib path to C
+            let path_str = path.map(|p| {
+                p.as_ref()
+                    .to_str()
+                    .expect("Path can't be represented as UTF-8")
+                    .to_owned()
+            });
+            if path_str.starts_with("\\\\?\\") {
```
Comment (sdroege) [inline src/svg.rs]: This and the following lines need to stay inside the `map` closure. You still have an `Option` here. Later you need to also get the pointer of the whole thing

File: src/svg.rs [hunk 5]
```diff
@@ -33,13 +35,39 @@ impl SvgVersion {
...
+        height: f64,
+        path: Option<P>,
+    ) -> Result<SvgSurface, Status> {
+        #[cfg(not(windows))]
+        let path = {
+            let path = path.map(|p| {
+                CString::new(p.as_ref().as_os_str().as_bytes())
+                    .expect("Invalid path with NULL bytes")
+            });
+            path.map(|p| p.as_ptr()).unwrap_or(ptr::null())
```
Comment (sdroege) [inline src/svg.rs]: The path will be freed once this block is done, and then you store a dangling pointer in path in the outer scope. Just return the `Option<CString>` from this block, and do the `as_ptr()` part for both Windows/UNIX just before calling the C function.

**Comment by sdroege** [comment]
Looks good for real now :)

**Comment by GuillaumeGomez** [comment]
Then I'll merge for real once CI passed. :laughing:

**Comment by EPashkin** [comment]
@bilelmoussaoui Thanks, 👍

**Comment by EPashkin** [comment]
@sdroege, @GuillaumeGomez Maybe we need to make these function public someday
https://github.com/gtk-rs/glib/blob/master/src/translate.rs#L422-L486

**Comment by GuillaumeGomez** [comment]
@EPashkin Oh indeed! Do you mind opening an issue for it please? So it can be done for the next release. :)

**Comment by sdroege** [comment]
That would not help here as otherwise cairo would have to always depend on glib

**Comment by EPashkin** [comment]
@sdroege Agree about this case, but IMHO here also better add separate `path_to_c` and use with `map` to make code more clear

**Comment by bilelmoussaoui** [comment]
> @sdroege Agree about this case, but IMHO here also better add separate `path_to_c` and use with `map` to make code more clear

I don't see a valid reason to do that as the function won't be used anywhere else but I agree it makes that part of code less readable. 

The failing tests are not related to this change I think.

**Comment by EPashkin** [comment]
@bilelmoussaoui Why you think that it not related? 
```
failures:

---- svg::test::file stdout ----

thread 'svg::test::file' panicked at 'called `Result::unwrap()` on an `Err` value: "error while writing to output stream"', src/libcore/result.rs:1189:5

failures:

    svg::test::file

test result: FAILED. 39 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by EPashkin** [comment]
IMHO test with `None` also will be good

**Comment by bilelmoussaoui** [comment]
Because I had that issue once, tried to reproduce it and it was gone the next time I ran the tests. Yeah a test with None will be good. Will add that.

File: src/svg.rs [hunk 6]
```diff
@@ -132,7 +158,7 @@ mod test {
     #[test]
     #[cfg(unix)]
     fn file() {
-        let surface = SvgSurface::new(100., 100., "/dev/null").unwrap();
+        let surface = SvgSurface::new(100., 100., Some("/dev/null")).unwrap();
```
Comment (EPashkin) [inline src/svg.rs]: Seems this test need be just `None`, and `Some` need use `tempfile`

Comment (bilelmoussaoui) [reply src/svg.rs]: Yeah, this is probably the reason why it fails.

File: src/svg.rs [hunk 7]
```diff
@@ -130,9 +156,16 @@ mod test {
...
+    fn without_file() {
+        let surface = SvgSurface::new(100., 100., None::<&Path>).unwrap();
+        draw(&surface);
+        surface.finish();
+    }
+
+    #[test]
     fn file() {
-        let surface = SvgSurface::new(100., 100., "/dev/null").unwrap();
+        let file = NamedTempFile::new().expect("tempfile failed");
```
Comment (EPashkin) [inline src/svg.rs]: Why not `tempfile()` ?

**Comment by sdroege** [comment]
Still fails to write the output file, on Windows and macOS but not on Linux.

**Comment by EPashkin** [comment]
On Windows `SvgSurface::new` not work even for direct path

**Comment by EPashkin** [comment]
Old version works on `let surface = SvgSurface::new(100., 100., "_.tmp").unwrap();`,
I try to find problem

File: src/svg.rs [hunk 8]
```diff
@@ -33,13 +35,37 @@ impl SvgVersion {
...
+                CString::new(path_str.as_bytes())
+            }
+            .expect("Invalid path with NUL bytes")
+        });
 
         unsafe {
             Ok(Self(Surface::from_raw_full(
-                ffi::cairo_svg_surface_create(path.as_ptr(), width, height),
+                ffi::cairo_svg_surface_create(
+                    path.map(|p| p.as_ptr()).unwrap_or(ptr::null()),
```
Comment (EPashkin) [inline src/svg.rs]: Windows works with `path.as_ref().map(|p| p.as_ptr()).unwrap_or(ptr::null()),`

Comment (sdroege) [reply src/svg.rs]: Ok yeah, not doing `as_ref()` will consume the `CString` here and we have a dangling pointer again. Not just for Windows.

File: src/svg.rs (hunk 7 above)
Comment (bilelmoussaoui) [reply src/svg.rs]: Because I couldn't find a way to get a `Path` from `tempfile()`

Comment (EPashkin) [reply src/svg.rs]: Your right

**Comment by EPashkin** [comment]
CI passed

**Comment by sdroege** [comment]
Great, let's get it in then :)

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**